src/errors.rs

```diff
@@ -1,79 +1,85 @@
-        InvalidCargoConfig {
-            description("Invalid cargo config")
-        }
-        /// Unable to find the source specified by 'replace-with'
-        NoSuchSourceFound(name: String) {
-            description("Unable to find the source specified by 'replace-with'")
-            display("The source '{}' could not be found", name)
-        }
-        /// Unable to find the specified registry
-        NoSuchRegistryFound(name: String) {
-            display("The registry '{}' could not be found", name)
-        }
-        /// Failed to parse a version for a dependency
-        ParseVersion(version: String, dep: String) {
-            description("Failed to parse a version for a dependency")
-            display("The version `{}` for the dependency `{}` couldn't be parsed", version, dep)
-        }
-    }
+/// Main error type
+#[derive(Debug, ThisError)]
+pub enum Error {
+    // foreign links
+    /// An error from the std::io module
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+    ///  An error from the git2 crate
+    #[error(transparent)]
+    Git(#[from] git2::Error),
+    /// An error from the cargo_metadata crate
+    #[error(transparent)]
+    CargoMetadata(#[from] cargo_metadata::Error),
+    /// An error from the toml_edit crate
+    #[error(transparent)]
+    TomlEditParse(#[from] toml_edit::TomlError),
+    /// A ReqParseError from the semver crate
+    #[error(transparent)]
+    SemVerParse(#[from] semver::ReqParseError),
+    /// A SemVerError from the semver crate
+    #[error(transparent)]
+    SemVer(#[from] semver::SemVerError),
+    /// An error from the reqwest crate
+    #[error(transparent)]
+    Reqwest(#[from] reqwest::Error),
+    /// An error from the serde_json crate
+    #[error(transparent)]
+    Json(#[from] serde_json::Error),
+    /// An error from converting bytes to a utf8 string
+    #[error(transparent)]
+    FromUtf8(#[from] std::string::FromUtf8Error),
+
+    /// A value was expected but we got `None` instead
+    #[error("Expected a value")]
+    ExpectedValue,
+
+    /// Given crate name is empty
+    #[error("Found empty crate name")]
+    EmptyCrateName,
+
+    /// No versions available
+    #[error("No available versions exist. Either all were yanked or only prerelease versions exist. Trying with the --allow-prerelease flag might solve the issue.")]
+    NoVersionsAvailable,
+
+    /// A resource was invalid
+    #[error("Invalid: {0}")]
+    Invalid(String),
+
+    /// Found a workspace manifest when expecting a normal manifest
+    #[error("Found virtual manifest, but this command requires running against an actual package in this workspace.")]
+    UnexpectedRootManifest,
+
+    /// Unable to find the specified resource
+    #[error("The resource `{0}` could not be found.")]
+    NotFound(String),
+
+    /// Unable to find the specified resource
+    #[error("The resource `{0}` could not be found in `{1}`.")]
+    NotFoundIn(String, String),
```

Comment:
these to also look similar

Reply:
I split them up because the second error message is helpful in giving a location for the error, i.e. it ends up looking like "the resource `serde` could not be found in `dev-dependencies`", and I liked having that location information.